File: src/main/java/destiny/core/calendar/LngLatDecorator.java
/**
 * Created by smallufo on 2017-05-05.
 */
package destiny.core.calendar;

import com.google.common.collect.ImmutableMap;
import destiny.tools.Decorator;
import destiny.tools.LocaleUtils;
import org.jetbrains.annotations.NotNull;

import java.util.Locale;

/**
 * 純粹只有經緯度，以及時區
 */
public class LngLatDecorator {

  private final static ImmutableMap<Locale , Decorator<Location>> implMap = new ImmutableMap.Builder<Locale , Decorator<Location>>()
    .put(Locale.TAIWAN , new LngLatDecoratorTaiwan())
    .put(Locale.CHINA  , new LngLatDecoratorChina())
    .put(Locale.ENGLISH, new LngLatDecoratorEnglish())
    .build();

  @NotNull
  public static String getOutputString(Location location , Locale locale) {
    return implMap.get(
      LocaleUtils.getBestMatchingLocale(locale, implMap.keySet()).orElse((Locale) implMap.keySet().toArray()[0])
    ).getOutputString(location);
  }
}

File: src/main/java/destiny/core/calendar/LngLatDecorator.kt
/**
 * Created by smallufo on 2017-05-05.
 */
package destiny.core.calendar

import destiny.tools.Decorator
import destiny.tools.getOutputString
import java.text.DecimalFormat
import java.util.*

/**
 * 純粹只有經緯度，以及時區
 */
object LngLatDecorator {

  private val implMap = mapOf<Locale, Decorator<Location>>(
    Locale.TAIWAN to LngLatDecoratorTaiwan(),
    Locale.CHINA to LngLatDecoratorChina(),
    Locale.ENGLISH to LngLatDecoratorEnglish()
  )

  fun getOutputString(location: Location, locale: Locale): String {
    return implMap.getOutputString(location, locale)
  }
}


class LngLatDecoratorTaiwan : Decorator<Location> {

  internal var formatter = DecimalFormat("00.00")

  override fun getOutputString(value: Location): String {
    return with(StringBuilder()) {
      append(LngDecorator.getOutputString(value.longitude , Locale.TAIWAN))
      append(", ")
      append(LatDecorator.getOutputString(value.latitude , Locale.TAIWAN))
    }.toString()
  }
}

class LngLatDecoratorChina : Decorator<Location> {

  internal var formatter = DecimalFormat("00.00")

  override fun getOutputString(value: Location): String {
    return with(StringBuilder()) {
      append(LngDecorator.getOutputString(value.longitude , Locale.SIMPLIFIED_CHINESE))
      append(", ")
      append(LatDecorator.getOutputString(value.latitude , Locale.SIMPLIFIED_CHINESE))
    }.toString()
  }
}


class LngLatDecoratorEnglish : Decorator<Location> {

  override fun getOutputString(value: Location): String {

    return with(StringBuilder()) {
      append(LngDecorator.getOutputString(value.longitude , Locale.ENGLISH))
      append(", ")
      append(LatDecorator.getOutputString(value.latitude , Locale.ENGLISH))
    }.toString()
  }
}
